gpiozero/pins/pi.py

```diff
@@ -217,6 +247,98 @@ def _get_spi_class(self, shared, hardware):
+        :exc:`~gpiozero.I2CBadArgs`.
+        """
+        i2c_args, kwargs = self._extract_i2c_args(**i2c_args)
+        shared = bool(kwargs.pop('shared', False))
+        if kwargs:
+            raise I2CBadArgs(
+                'unrecognized keyword argument {arg}'.format(
+                    arg=kwargs.popitem()[0]))
+        try:
+            return self._get_i2c_class(shared, hardware=True)(
+                pin_factory=self, **i2c_args)
+        except Exception as e:
+            warnings.warn(
+                I2CSoftwareFallback(
+                    'failed to initialize hardware I2C, falling back to '
+                    'software (error was: {e!s})'.format(e=e)))
+        return self._get_i2c_class(shared, hardware=False)(
+            pin_factory=self, **i2c_args)
+    
+    def _extract_i2c_args(self, **kwargs):
+        """
+        Given a set of keyword arguments, splits it into those relevant to I2C
+        implementations and all the rest. I2C arguments are augmented with
+        defaults and converted into the pin format (from the port format) if
+        necessary.
+
+        Returns a tuple of ``(i2c_args, other_args)``.
+        """
+        try:
+            device = kwargs.pop("device")
+        except KeyError:
+            raise I2CBadArgs(
+                'device address is not specified as the *device* keyword'
+            )
+        dev_defaults = {
+            'port': 1,
+            }
+        default_hw = I2C_HARDWARE_PINS[dev_defaults['port']]
+        pin_defaults = {
+            'data_pin':  default_hw['data'],
+            'clock_pin':  default_hw['clock'],
+            }
+        i2c_args = {
+            key: value for (key, value) in kwargs.items()
+            if key in pin_defaults or key in dev_defaults
+            }
+        kwargs = {
+            key: value for (key, value) in kwargs.items()
+            if key not in i2c_args
+            }
+        if not i2c_args:
+            i2c_args = pin_defaults
+        elif set(i2c_args) <= set(pin_defaults):
+            i2c_args = {
+                key: None if i2c_args.get(key, default) is None else
+                    self.pi_info.to_gpio(i2c_args.get(key, default))
+                for key, default in pin_defaults.items()
+                }
+        elif set(i2c_args) <= set(dev_defaults):
+            try:
+                selected_hw = I2C_HARDWARE_PINS[i2c_args['port']]
+            except KeyError:
+                raise I2CBadArgs(
+                    'port {i2c_args[port]} is not a valid I2C port'.format(
+                        i2c_args=i2c_args))
+            i2c_args = {
+                "data_pin": selected_hw["data"],
+                "clock_pin": selected_hw["clock"]
+                }
+        else:
+            raise I2CBadArgs(
+                'you must either specify port and device, or clock_pin, '
+                'mosi_pin, miso_pin, and select_pin; combinations of the two '
```

Comment:
I guess the part here talking about mosi_pin and miso_pin etc. needs to be changed :wink:

Reply:
Yup, sloppy copying 🙈